File: Array/957. Prison Cells After N Days.cpp
class Solution {
public:
    vector<int> prisonAfterNDays(vector<int>& res, int n) {
        vector<int> tmp(8);
        vector<vector<int>> seen;
        while (n--) {
            for (int i = 1; i < 7; i++) {
                tmp[i] = res[i - 1] == res[i + 1];
            }
            if (seen.size() && seen.front() == tmp) return seen[n % seen.size()]; 
            else seen.push_back(tmp);
            res = tmp;
        }
        return res;
    }
};


class Solution {
public:
    vector<int> prisonAfterNDays(vector<int>& cells, int n) {
        vector<int>ans(8);
        ans=cells;
        int m=(n-1)%14;
        while(m>=0)
        {
            for(int i=0;i<=7;i++)
            {   
                if(i==0|| i==7)
                    ans[i]=0;
                else if(cells[i-1]== cells[i+1])
                    ans[i]=1;
                else
                    ans[i]=0;
            }
            m--;
            for(int i=0;i<8;i++)
            {
                cells[i]=ans[i];
            }
        }
        return ans;
    }
};




class Solution {
public:
    vector<int> prisonAfterNDays(vector<int>& cells, int N) {
        for (N = (N - 1) % 14 + 1; N > 0; --N) {
            vector<int> cells2(8, 0);
            for (int i = 1; i < 7; ++i)
                cells2[i] = cells[i - 1] == cells[i + 1] ? 1 : 0;
            cells = cells2;
        }
        return cells;
    }
};

File: Array/957. Prison Cells After N Days.py
class Solution:
    def prisonAfterNDays(self, cells: List[int], n: int) -> List[int]:
        def get_next(cells):
            ret = [0]
            for i in range(1, len(cells)-1):
                ret.append(int(cells[i-1] == cells[i+1]))
            ret.append(0)
            return ret

        seen = dict()
        is_fast_forward = False

        while n > 0:
            if not is_fast_forward:
                key = tuple(cells)
                if key in seen:
                    n = n % (seen[key] -  n)
                    is_fast_forward = True
                else:
                    seen[key] = n
            if n > 0:
                n -= 1
                cells = get_next(cells)
        return cells

class Solution:
    def prisonAfterNDays(self, cells: List[int], N: int) -> List[int]:

        seen = dict()
        is_fast_forwarded = False

        # step 1). convert the cells to bitmap
        state_bitmap = 0x0
        for cell in cells:
            state_bitmap <<= 1
            state_bitmap = (state_bitmap | cell)

        # step 2). run the simulation with hashmap
        while N > 0:
            if not is_fast_forwarded:
                if state_bitmap in seen:
                    # the length of the cycle is seen[state_key] - N 
                    N %= seen[state_bitmap] - N
                    is_fast_forwarded = True
                else:
                    seen[state_bitmap] = N
            # If there are still some steps remaining,
            # with or without the fast-forwarding.
            if N > 0:
                N -= 1
                state_bitmap = self.nextDay(state_bitmap)

        # step 3). convert the bitmap back to the state cells
        ret = []
        for i in range(len(cells)):
            ret.append(state_bitmap & 0x1)
            state_bitmap = state_bitmap >> 1

        return reversed(ret)


    def nextDay(self, state_bitmap: int):
        state_bitmap = ~ (state_bitmap << 1) ^ (state_bitmap >> 1)
        state_bitmap = state_bitmap & 0x7e  # set head and tail to zero
        return state_bitmap



class Solution:
    def prisonAfterNDays(self, cells: List[int], N: int) -> List[int]:

        seen = {str(cells): N}
        while N:
            seen.setdefault(str(cells), N)
            N -= 1
            cells = [0] + [cells[i - 1] ^ cells[i + 1] ^ 1 for i in range(1, 7)] + [0]
            if str(cells) in seen:
                N %= seen[str(cells)] - N
        return cells